httpx/concurrency/asyncio.py

```diff
@@ -232,7 +237,7 @@ def loop(self) -> asyncio.AbstractEventLoop:
     ) -> BaseTCPStream:
         try:
             stream_reader, stream_writer = await asyncio.wait_for(  # type: ignore
-                asyncio.open_connection(hostname, port, ssl=ssl_context),
+                self._open_connection(hostname, port, ssl_context),
```

Comment:
Is there a way we could generalize the backend interface so that it exposes a way to open a unix socket, rather than create a separate backend?

This way, users won't need to use a specific backend, and we can later look at adding this piece of functionality to the trio backend, too.

I'm thinking, maybe a new `.open_unix_socket()` method similar to the current `.open_tcp_socket()`, and then make the switch between the two in `HTTPConnection.connect()` (if I remember correctly that that's where we create a stream instance).

Reply:
Great feedback! I was thinking the same but wasn't sure if stream dispatching would fit that high in the stack.

If we dispatch between tcp and unix connection in `HttpConnection`, we need to determine what makes it select one or the other.

One example could be the `origin.scheme` being `unix` or `http+unix` to favour the unix socket route. But that affects other bits too much maybe, and also the path to the socket file needs to be passed some way, e.g. as `origin.host` which in turn hides any wanted real hostname to be passed through the unix socket.

For now, it might be better to add the unix socket path as an optional argument to the base class `ConcurrencyBackend` and implement the dispaching there in new `.open_stream()` method, which in turn calls your suggested `.open_tcp_stream()` / `.open_unix_stream()`.

That way both `AsyncioBackend`, `TrioBackend` etc gets support for the socket file feature, but one needs to instantiate the backend with the socket file arg as my current implementation in the separate backend does.

Reply:
Hmm. Well, if we put aside the whole concurrency backend aspect (which is only us trying to abstract I/O functionality so that we aren't particularly tied to any async library), *maybe* there's a way to, as @tomchristie suggested, to implement this as a separate, very narrow-focused dispatcher.

It's clearer to me why it should be doable if we look at [requests-unixsocket](https://pypi.org/project/requests-unixsocket/), which provides unix sockets in the form of a `Session` subclass that uses its own adapter (a concept somewhat similar to dispatchers in HTTPX). The implementation is based on HTTP connection and connection pool subclasses. In theory we should be able to take the same path in HTTPX.

And then in terms of API, maybe aim at something like…

```python
import httpx
from httpx_unixsocket import UnixSocketAsyncDispatcher

dispatch = UnixSocketAsyncDispatcher("/var/run/docker.sock")
async with httpx.AsyncClient(dispatch=dispatch) as client:
    r = await client.get("http+unix://docker/version")
```